### PR TITLE
config: stats flush via admin

### LINKED ISSUE #14007: Interval Latched Histograms
Histogram: Support flushing histograms on /stats instead of on a timer

From the documentation : "_The output for each quantile will be in the form of (interval,cumulative) where interval value represents the summary since last flush interval and cumulative value represents the summary since the start of Envoy instance_"

The cumulative metrics doesn't do a good job in reflecting the current latency profile of a running instance. The _interval_ value relies on the stats timer which is sufficient if we rely on that timer when exporting metrics. For systems that collect metrics from an adhoc process on-demand via an endpoint like `/stats`, can we support latching the histograms until the stats endpoint is queried to fully capture an interval? This is something I can work on contributing.

### PR DESCRIPTION
**Pull request by hamdiallam**
Commit Message: Add the ability to flush stats only when queried for.
Additional Description: Fixes #14007. Add new field `stats_flush` where either optionally `stats_flush_on_admin`  or `stats_flush_interval` is specified (in v4). In the v3 api, we introduce stats_flush and annotate stats_flush_interval with the oneof_promotion annotation.

Risk Level: Low
Testing: `server_test.cc` already accounts for manually calling `flushStats`. Add another test where stats flushing is driven through the admin interface.
Docs Changes: API documentation.
Release Notes: Added config change
Platform Specific Features: none

### CODE REVIEW

File: test/server/server_test.cc
Comment (hamdiallam) [inline test/server/server_test.cc:527]: Must we run `BLOCK` on the dispatcher after this?

**Comment by adisuissa** [comment]
Seems like PR #14139 also attempts to fix the same issue.

**Comment by ramaraochavali** [comment]
@hamdiallam Sorry, I pushed the PR some time back to fix that https://github.com/envoyproxy/envoy/pull/14139. But if you think you would want to continue with this - I am happy to close my PR

File: api/envoy/config/bootstrap/v2/bootstrap.proto
Comment (jmarantz) [inline api/envoy/config/bootstrap/v2/bootstrap.proto:107]: you can't retroactively change the v2 API.

File: source/server/configuration_impl.h
Comment (jmarantz) [inline source/server/configuration_impl.h:124]: it doesn't look like this gets initialized on all paths. Maybe declare it as
```
stats_flush_on_admin_{false}
```
to be sure?

File: test/server/server_test.cc
Comment (jmarantz) [inline test/server/server_test.cc:524]: just split this into two tests so we continue to directly test server_->flushStats()

**Comment by hamdiallam** [comment]
> @hamdiallam Sorry, I pushed the PR some time back to fix that #14139. But if you think you would want to continue with this - I am happy to close my PR

I'd be great to make the contribution! 😄 Took this as an opportunity to familiarize myself more with envoy's internals

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (adisuissa) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:168]: As @jmarantz mentioned v2 is frozen and should not be changed.
In v3 changing the `stats_flush_interval` to oneof will break [backwards compatibility](https://github.com/envoyproxy/envoy/blob/master/api/API_VERSIONING.md#backwards-compatibility).
You should be able to annotate this as oneof_promotion (see example [here](https://github.com/envoyproxy/envoy/blob/de222c3acf1fbf124ca44f50495fc4b8e9aba525/tools/testdata/protoxform/envoy/v2/oneof.proto) and [here](https://github.com/envoyproxy/envoy/blob/de222c3acf1fbf124ca44f50495fc4b8e9aba525/tools/testdata/protoxform/envoy/v2/oneof.proto.next_major_version_candidate.gold)) in case you wish to make it a oneof in the future.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (adisuissa) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:181]: Not an ideal solution, but using this field can supersede the interval usage.
Another option is to use both interval and admin flush concurrently.

**Comment by ramaraochavali** [comment]
> I'd be great to make the contribution! 😄 Took this as an opportunity to familiarize myself more with envoy's internals

Sure. Great. I would close my PR

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Here are some API comments.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (hamdiallam) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:181]: How would they be used concurrently? By setting the `stats_flush_on_admin`, we disable the timer. If they can be set concurrently, is there a way we can indicate to not flush on a timer? Maybe a duration of 0s?

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (hamdiallam) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Sounds good. Went ahead and used the oneof_promotion annoatation

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (adisuissa) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:184]: I think the current suggestion makes sense in terms of API compliance.
Can you please also update `envoy/docs/root/operations/admin.rst`?

Comment (hamdiallam) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:184]: lmk how it looks

Comment (adisuissa) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:184]: Thanks!

**Comment by hamdiallam** [comment]
Is there a way to retrigger the "CodeQL-Build" check. Seems unrelated to the small changes in the last commit

**Comment by jmarantz** [comment]
Don't worry about CodeQL for now; it's not required for merging. But I'll let Adi finish the review and then I'll do a pass.

File: docs/root/operations/admin.rst
Comment (adisuissa) [inline docs/root/operations/admin.rst:412]: I think that it would be more clear if the sentences are reordered, and that no timer will be used is explicitly stated:

```suggestion
  represents the summary since last flush. By default, a timer is setup to flush in intervals defined by :ref:`stats_flush_interval <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.stats_flush_interval>`, defaulting to 5 seconds.
If :ref:`stats_flush_on_admin <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.stats_flush_on_admin>` is specified, stats are flushed when this endpoint is queried and a timer will not be used. The cumulative value represents the summary since the start of Envoy instance.
```

File: source/server/configuration_impl.cc
Comment (adisuissa) [inline source/server/configuration_impl.cc:90]: nit: no need for curly brackets in case.

File: source/server/configuration_impl.cc
Comment (adisuissa) [inline source/server/configuration_impl.cc:95]: If nothing is done here, why not use an if clause instead of switch?

File: source/server/server.cc
Comment (adisuissa) [inline source/server/server.cc:548]: This comment needs to be updated/moved into the if clause.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Just a few minor comments.

File: docs/root/operations/admin.rst
Comment (hamdiallam) [reply docs/root/operations/admin.rst]: I agree, that sounds better!

File: source/server/configuration_impl.cc
Comment (hamdiallam) [reply source/server/configuration_impl.cc]: if statement makes the most sense here

**Comment by adisuissa** [comment]
LGTM.
@jmarantz Can you please take a look.

File: source/server/server.cc
Comment (jmarantz) [inline source/server/server.cc:554]: Can initialize() be called twice? Because if so I think we'll be adding the sinks a second time in lines 544-546 above. And i fit can only be called once I think stat_flush_timer_ cannot be populated so there should be no need to reset it, though you could ASSERT that.

File: test/server/server_test.cc
Comment (jmarantz) [inline test/server/server_test.cc:581]: EXPECT_TRUE

File: test/server/server_test.cc
Comment (jmarantz) [inline test/server/server_test.cc:585]: nit: capture the counter in a local rather than calling findCounter twice?

**Review by jmarantz — APPROVED** [review]
LGTM; modulo some minor comments.

I think @mattklein123 should do a senior-approver pass.

**Comment by jmarantz** [comment]
an @envoyproxy/api-shepherds will also need to approve, and one question I had was whether even the v3 change was compatible enough. Wire-compatible is not enough FYI, but I'll let an API expert explain in more detail.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (htuch) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:184]: Doing final API review here, @hamdiallam can you comment on the use case here? I'm trying to figure out if we ever would want an option for flushing with timer _and_ with admin endpoint access?

Comment (hamdiallam) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:184]: We don't ever want an option for flushing with the timer and admin endpoint access. For backwards compatibility, we can't use `oneof` to enforce that outside of adding the `oneof_promotion` annotation which in the v4 api, becomes a `oneof`. 

If both are specified in v3, admin endpoint access takes precedence. Perhaps a warning log entry that both were specified?

File: source/server/server.cc
Comment (hamdiallam) [reply source/server/server.cc:554]: This makes sense to me. The `reset` was more of a sanity thing but the ASSERT makes more sense

**Comment by hamdiallam** [comment]
> an @envoyproxy/api-shepherds will also need to approve, and one question I had was whether even the v3 change was compatible enough. Wire-compatible is not enough FYI, but I'll let an API expert explain in more detail.

Yep! This makes sense. I wasn't aware `oneof` wasn't backwards compatible and updated to use the annotation so preserve compatibility while improving the api for v4. Curious to know if I've missed something

Thanks for the review!

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (mattklein123) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:184]: > If both are specified in v3, admin endpoint access takes precedence. Perhaps a warning log entry that both were specified?

I would make this a config failure similar to what would be done if it were a oneof. This is what we do elsewhere.

File: include/envoy/server/configuration.h
Comment (mattklein123) [inline include/envoy/server/configuration.h:95]: Instead of polluting the main interfaces with various stats config, can we make a new interface for `StatsConfig` and just pass that around? Same below.

File: source/server/server.cc
Comment (mattklein123) [inline source/server/server.cc:555]: nit: I don't think this else clause adds much value. I would remove it.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with some small comments, thank you!

/wait

File: include/envoy/server/configuration.h
Comment (hamdiallam) [reply include/envoy/server/configuration.h:95]: This makes sense to me. Should we reflect the same in the v4 api?

Comment (mattklein123) [reply include/envoy/server/configuration.h:95]: I don't think I would bother with an API change here, just internal code change.

Comment (hamdiallam) [reply include/envoy/server/configuration.h]: sounds good

**Comment by hamdiallam** [comment]
> LGTM with some small comments, thank you!
> 
> /wait

Thanks! As c++ novice lmk if I've missed anything

**Comment by mattklein123** [comment]
Can you check format CI?

/wait

**Comment by mattklein123** [comment]
Needs main merge, thanks.

/wait

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (mattklein123) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:182]: Can you update the docs for this field as well as the other one to say that only one of them can be set?

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:61]: nit: ref link to stats endpoint

File: include/envoy/server/configuration.h
Comment (mattklein123) [inline include/envoy/server/configuration.h:75]: Make this method return a const reference to the list, and also a const method in general. Then, you can populate the list directly into the StatsConfigImpl via an impl specific method.

File: source/server/configuration_impl.cc
Comment (mattklein123) [inline source/server/configuration_impl.cc:61]: Please add a test for this case.

File: source/server/server.cc
Comment (mattklein123) [inline source/server/server.cc:558]:
```suggestion
  auto& stats_config = config_.statsConfig();
```

File: source/server/server.cc
Comment (mattklein123) [inline source/server/server.cc:241]:
```suggestion
  auto& stats_config = config_.statsConfig();
```

File: test/mocks/server/instance.h
Comment (mattklein123) [inline test/mocks/server/instance.h:133]: Define constructor here and impl in cc file also

File: test/server/configuration_impl_test.cc
Comment (mattklein123) [inline test/server/configuration_impl_test.cc:166]: nit: start with capital, end with period, etc. Same elsewhere.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, nice work. LGTM with some small comments.

/wait

File: source/server/configuration_impl.cc
Comment (hamdiallam) [reply source/server/configuration_impl.cc:61]: There's a test case for this in configuration_impl_test, `IntervalAndAdminFlush`.

File: test/server/configuration_impl_test.cc
Comment (mattklein123) [inline test/server/configuration_impl_test.cc:187]: Use THROW_WITH_MESSAGE and check the actual message here.

File: source/server/configuration_impl.cc
Comment (hamdiallam) [inline source/server/configuration_impl.cc:112]: @mattklein123 is it fine to down case here for the purposes of the impl specific `addSink`. Given createStatsSink takes a ServerFactoryContext as an argument, it has access to `stats_config` which is typed to `Configuration::StatsConfig`.

File: source/server/configuration_impl.h
Comment (hamdiallam) [inline source/server/configuration_impl.h:57]: Not sure of best practices of when to inline the definition. This seemed appropriate

**Comment by hamdiallam** [comment]
> Thanks, nice work. LGTM with some small comments.
> 
> /wait

Thanks for the feedback

File: source/server/configuration_impl.cc
Comment (mattklein123) [reply source/server/configuration_impl.cc:112]: Just make stats_config_ the impl type and you don't need the cast. It can be returned as the interface type.

/wait

Comment (hamdiallam) [reply source/server/configuration_impl.cc]: ah yes of course

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (mattklein123) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:169]: nit: space after period

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:61]: Also ref link to the new field that configures this.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with a few small comments.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by jmarantz** [comment]
I realize I reviewed this PR, but why did we make flushOnAdmin default to true?

It basically means we need to do a full stats sink any time you access any /stats endpoint, which may double the cost or worse. The functionality seems useful but why did we decide to make this the default?

**Comment by hamdiallam** [comment]
It's not set to true by default, the value this config property can have is strictly "true"

Otherwise the default [is an in interval](https://github.com/envoyproxy/envoy/pull/14156/files#diff-998fda72180293e9c2eaf07d3ead906aba6998220cbcd9f10f63bc283f5c6119R64) for `stats_flush_interval` which preserves old behavior.  You can see the default value for `flush_on_admin_` is false

**Comment by jmarantz** [comment]
Hm I got confused -- I thought I had seen evidence in the debugger that it was getting into the stats-sinks in response to a /stats request, with no special command-line args.

Never mind then...I'll ping again if I notice this.

**Pull request merged**
